pg_search/src/postgres/customscan/hook.rs

```diff
@@ -295,9 +406,23 @@ unsafe extern "C-unwind" fn paradedb_planner_hook(
         if query_has_window_func_nodes(parse) {
             let has_paradedb_agg = query_has_paradedb_agg(parse);
             if query_is_topn(parse) {
-                if has_paradedb_agg || query_has_search_operator(parse) {
-                    // Extract and replace window functions recursively (including subqueries)
-                    replace_windowfuncs_recursively(parse);
+                let has_search_operator = query_has_search_operator(parse);
+                if has_paradedb_agg || has_search_operator {
+                    // Before replacing window functions, check if WHERE clause can be handled
+                    if !can_handle_where_clause(parse) {
+                        if has_paradedb_agg {
+                            // pdb.agg() requires that we handle the query, but we can't guarantee all WHERE predicates can be handled
+                            pgrx::error!(
+                                "pdb.agg() window functions cannot be used with this WHERE clause because some predicates may not be pushable to the index. \
+                                 To fix this, enable 'SET paradedb.enable_filter_pushdown = on' to allow filtering on all fields. \
+                                 Alternatively, ensure all WHERE predicates use indexed fields or remove non-indexed predicates."
+                            );
+                        }
+                        // Otherwise, don't replace window functions - let PostgreSQL handle them
+                    } else {
+                        // Extract and replace window functions recursively (including subqueries)
+                        replace_windowfuncs_recursively(parse);
+                    }
                 }
             } else if has_paradedb_agg {
```

Comment:
This block is very, very nested, which makes it hard to review.

Would be good to flatten it as much as possible, possibly by extracting the detection of whether we _can_ replace window functions into a function, and then having that function early-return.

Reply:
Refactored it and reduced nesting.